### DISCUSSION TITLE
Towards a polyphonic modulation system

### DISCUSSION THREAD

**Discussion by RobinSchmidt** [RobinSchmidt/RS-MET#312] [Ideas]
i want to finally make the modulation system polyphony-ready. i have wanted to implement my dream-synth since a long time. the basic architecture should be semi modular, with 4 source-slots that can be equipped with arbitrary algorithms (oscillators, sample players, FM-sources, etc.) and then be mixed via a 2D vector-mixer. the result of that should go into a dual filter that also has 2 slots into which arbitrary filter algorithms can be plugged in and also has a vector mixer that (somehow, not yet sure how) morphs between various configurations of the 2 filters (serial, parallel, etc.). all parameters should be modulatable by an arbitrary number of modulation sources - which can also be plugged in into slots. all of that should work polyphonically. 

the synth's architecture is basically very similar to straightliner's, but taken to the next level. instead of 4 oscillators, there are 4 arbitrary sources, instead of the single filter, there's a dual filter, where actually each "filter" can also be some other type of effect, and the 3 hardwired breakpoint-modulators are replaced by an arbitrary number of routable modulators of various kinds

...soo i started thinking about, how i would implement the underlying polyphonic modulation system and realized that maybe a better idea would be to not start from scratch but rather extend the existing system to facilitate polyphony. i would then actually not really need to implement the synthesizer as such but could use the existing infrastructure of ToolChain to turn ToolChain *itself* into the synthesizer i was just describing. i want to implement the QuadSource and the DualFilter as separate modules that can be loaded into 2 subsequent slots in ToolChain. then i could modulate their parameters by loading modulators into other slots (and then maybe have for convenience a source of multiple modulation signals that has itself slots, so as to not have to fill too many of ToolChain's primary slots with all the modulators). ....and all of that calls for a polyphonic modulation system

**Comment by denisreview** [comment]
(reposting into the right thread)

Hi Robin,

Speaking of synths and polyphony, I've been obsessed with an idea to integrate into the daw a built-in synth, with powerful modulation capabilities and with a cinematic "feel" to it, close to what the Absynth (now discontinued by NI) has been able to produce. Are you familiar with Absynth & its features? I think it was discontinued, because it was too complex for the general public to get handle of, and to understand its endless GUI. Would be cool to create something similar to Absynth, but with a simpler GUI, signal scheme, and more friendly overall.

So, the main question is, can your current "dream synth" ideas be applied somehow here? Anyhow, just my 2 cents, from what I'd consider my own kind of "dream synth". What are the main goals/benefits of polyphony, anyway?

**Reply by RobinSchmidt** [reply]
i'm not very familiar with absynth. i dabbled a bit with it back then in ...i think 2003 or 2004... when i had an internship at NI. but i can't really say that i know the synth very well. i think, complexity is not necessarily a bad thing as long as it is optional. for a novice user, it should be all hidden and everything should just work out of the box (i.e. the default behavior should be meaningful) an advanced user should be able to dig deeper and customize all sorts of things. so, what would your dream synth look like? what general architecture do you have in mind?

what are the main goals of polyphony, anyway? ...well, obviously, to play multiple notes at once. i'm not sure, if i understand that question right. ...edit: or maybe you are confusing "polyphonic" with "multitimbral" here?. i think  the "multitimbral" feature does indeed not make very much sense in the context of software instruments where you can just fire up multiple instances, if needed - which you obviously can't do with hardware, which is why there, it is still a thing. in software instruments, i think, the multitimbrality feature would just add unnecessary complexity

**Comment by denisreview** [comment]
Yes, probably I was confusing the two here. But speaking of playing multiple notes at once, isn't that what Straightliner was already perfectly capable of, with the Voices setting? Or being truly polyphonic just makes the number of voices unlimited?

Speaking of my dream synth, I don't have a very clear idea yet, but I'd like it to have some powerful stereoization features, allowing to experiment wildly with sound stereobase and behaviour in 3d space. Also, allowing enrichment of sound with nonlinear distortions, & similar stuff, emulating analog circuits, or some weird physical models. And another thing - the ability for the synth to generate multitudes of random patches, so you could have lots of interesting "raw" patches, with further ability to refine and get interesting new results from this material. 

But it's all just in the realm of deas now, ofc. Absynth is one synth that was kind of inspiring for me with its capabilities & sounds & looks. If I could create something close in the future, as built-in part of the DAW, that would be my dream (or one of my dreams) come true, I think.

**Reply by RobinSchmidt** [reply]
yes, straightliner does have polyphony but the way i have implemented it back then is not suitable for letting the user route the modulations at runtime. there, the polyphonic modulations are all hardwired in the code (and that in a rather heavy handed way). no, (i think) in polyphonic synths, the number of voices is always limited. sometimes to 16, sometimes to 32, sometimes even to 128, but i think, there always has to be some limit

stereoization in 3D space? are you talking about some sort HRTF processing here? or do you want to make a synth with surround output, like 5.1, 7.1, etc? that would actually sound like an appropriate thing to do for a "cinematic" synth

**Reply by denisreview** [reply]
No, I mean just normal stereo synth, with stereo output. No too complex processing, just making the sound as stereo as possible, using basic modulation with phase difference, microdelays, maybe some equalization? How many ways are there to manipulate two channels and making the sound more stereo-rich & interesting? Which aren't too cpu-heavy, so they could be adopted in a synth.
